---
title: ' RNAseq Analysis'
author: "Hernan Lorenzi"
date: "12/06/2023"
output:
  html_document:
    df_print: paged
---

#### Load libraries
```{r load libraries}
pacman::p_load(pheatmap,EnhancedVolcano,ggpubr,DESeq2,stringr,biomaRt,tidyverse,pcaExplorer,VennDiagram,clusterProfiler,GOSemSim,ggsci,viridis,ggrepel,RColorBrewer,msigdbr,cowplot,enrichplot,ggupset,broom,ggraph)

```

#### Define functions
```{r aux functions}

# Load auxyliary functions
source(file = "./01_aux_rnaseq_functions.R")

# Load enrichment functions
source(file = "./02_Gene_enrichment_functions.R")

```


#### Load read counts data
```{r Loading data}
all.star <- read.delim2("./data/read_counts.txt", sep = "\t", header = TRUE, row.names = 1, comment.char = c("#") )

format_star <- function(star_file){
  names(star_file) <- names(star_file) %>% 
    str_remove_all(pattern = "results.03map_reads.|.Aligned.sortedByCoord.out.bam") 
  return(star_file[6:ncol(star_file)])
}

# Format star counts file
all <- format_star(star_file = all.star)

# Make sure read counts are numeric and rounded to 0 decimals
all.tmp <- as.data.frame(lapply(all, function(x){ round(as.numeric(x), digits = 0)} ))
rownames(all.tmp) <- rownames(all)
all <- all.tmp

#Remove all zero rows
all <- remove_all_zero_rows(all, min_total_count = 0)

```

#### Make metadata table from 'all'
```{r}
# Load metadata
metadata <- read.delim2("./data/metadata.csv", sep = ",", header = TRUE, row.names = 1, comment.char = c("#") )

# sort all columns based on metadata row names
all <- all %>% select(rownames(metadata))

# Add total read counts and sample id columns to metadata
metadata$read_counts <- colSums(all)

# Add "Sample_name" as column in metadata
metadata$sample_name <- rownames(metadata)

# edit treatment column
metadata$treatment <- str_remove(metadata$treatment, pattern = "Grown at ")

# Kepp columns of interest
metadata <- metadata %>% select(c("genotype","treatment","read_counts","sample_name"))

# change label for mutant
metadata[metadata$genotype == "RpoD D445V mutant","genotype"] <- "mutant"

# Add column combining genotype and treatment
metadata$group <- paste(metadata$treatment,metadata$genotype, sep = "_")

```


#### Normalize data to TPMs to run some comparative analysis across samples
```{r}
# Using annotation version GRCm39 (current)
all.tpm <- normalize_by_TPM(counts.df = all, gene_length = select(all.star, c("Length")))
```

###########################################
# Analysis of expression data using DESeq2
###########################################

```{r deseq2}
# Convert metadata to factors
for (variable in c("genotype", "treatment","sample_name","group")){
  metadata[,variable] <- as.factor(str_replace_all(metadata[,variable], pattern = " ", replacement = "_"))  
}

```

# Analysis of Dataset

```{r Dataset one}
# Generate DESeq2 object for NS and ST condition ONLY. We could potentially add Read_counts as either a covariate or as a factor (by grouping Read counts into Low and High count bins). Ideally, this should be done adding Read_counts as a random effect using mixed models, that can only be applied in Limma package. Adding it as a fixed effect in the design formula might lower the power of the analysis.

dir.create(path = "./Plots", showWarnings = FALSE)

# Create DESeq object
dds.all <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~ group)

# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "sample_name", 
          y = "read_counts",
          x.text.angle = 90,
          fill = "group", 
          title = "Total read counts", 
          ylab = "Read count",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc") 
ggsave2("Plots/barplot_read_counts.pdf", plot = p)
print(p)
```


```{r}
# Normalize counts
vsd.one <- vst(dds.all, blind=FALSE)
rlog.one <- rlog(dds.all, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(counts(dds.all)) >= 20
dds.all <- dds.all[keep,]

# Calculate distances between samples
sampleDists <- dist(t(assay(vsd.one)))

# Plot inter-sample distances
old.par <- par(no.readonly=T)

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rlog.one$treatment, rlog.one$genotype, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
p.pheatmap <- pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

ggsave2(filename = "unsupervised_clustering_rnaseq_profile_20plus_reads.pdf", plot = p.pheatmap, path = "./Plots")
print(p.pheatmap)
```

```{r}
dds_res <- list()

dds_res <- dds.all #[ , dds.all$Tissue=="all_data"]

rlog_res <- list()
rlog_res <- rlog(dds_res, blind=FALSE)

# PCA
rlog.one <- rlog_res

# PC1 - PC2
PC_coords = c(1,2)
pcaData <- plotPCA_anycoord(object = rlog.one, intgroup=c("genotype","treatment"), returnData=TRUE, PC_coords = PC_coords)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords

p12 <- ggplot(data = pcaData, aes(x = PC1, y = PC2, color = genotype, shape = treatment)) +
  geom_point(size = 3) + 
  xlab(paste0("PC",PC_coords[1],": ", round(percentVar[1]), "% variance")) + 
  ylab(paste0("PC",PC_coords[2],": ", round(percentVar[2]), "% variance")) + 
  coord_fixed() + 
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + # Remove grid
  scale_color_brewer(type = "div", palette = "Set1") + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) 

ggsave(paste0("Plots/pca_PC",PC_coords[1],PC_coords[2],"_Group.pdf"), plot = p12)
print(p12)

# PC2 - PC3
PC_coords = c(2,3)
pcaData <- plotPCA_anycoord(object = rlog.one, intgroup=c("genotype","treatment"), returnData=TRUE, PC_coords = PC_coords)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords
p1 <- ggplot(data = pcaData, aes(x = PC1, y = PC2, color = genotype, shape = treatment)) +
  geom_point(size = 3) + 
  xlab(paste0("PC",PC_coords[1],": ", round(percentVar[1]), "% variance")) + 
  ylab(paste0("PC",PC_coords[2],": ", round(percentVar[2]), "% variance")) + 
  coord_fixed() + 
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + # Remove grid
  scale_color_brewer(type = "div", palette = "Set1") + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) 

print(p1)
ggsave(paste0("Plots/pca_PC",PC_coords[1],PC_coords[2],"_Group.pdf"), plot = p1)

# PC1 - PC3
PC_coords = c(1,3)
pcaData <- plotPCA_anycoord(object = rlog.one, intgroup=c("genotype","treatment"), returnData=TRUE, PC_coords = PC_coords)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords

p13 <- ggplot(data = pcaData, aes(x = PC1, y = PC2, color = genotype, shape = treatment)) +
  geom_point(size = 3) + 
  xlab(paste0("PC",PC_coords[1],": ", round(percentVar[1]), "% variance")) + 
  ylab(paste0("PC",PC_coords[2],": ", round(percentVar[2]), "% variance")) + 
  coord_fixed() + 
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + # Remove grid
  scale_color_brewer(type = "div", palette = "Set1") + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) 

print(p13)
ggsave(paste0("Plots/pca_PC",PC_coords[1],PC_coords[2],"_Group.pdf"), plot = p13)

```
PCA analysis shows that samples separate by genotype and treatment.

### resultsNames(dds)

#### Filtering out poorly-expressed genes (less than 20 reads across all samples) 
```{r filtering genes based on total counts of reads}

# Keep genes with at least 10 reads total across samples
keep <- rowSums(counts(dds_res)) >= 20
dds_res <- dds_res[keep,]

```


##### Using groups instead of interactions
```{r}
ensembl_to_symbol <- read.delim(file = "./data/gene_names.txt", col.names = c("Ensembl_ID","gene_name"), header = F)

# Save sorted files as a list
DE_results <- list()
geneids.DE <- list()

# Define function for processing and saving result tables
sort_and_write_res_table <- function(result_table, file_name){
  dir.create(path = "./DE", showWarnings = FALSE)
  # Sort genes by (padj)
  result_table_sorted <- result_table[order(result_table$padj, decreasing = FALSE),]
  # Add gene symbols
  gene_list <- rownames(result_table_sorted)
  symbol_list <- ensembl_to_symbol$gene_name[match(gene_list, ensembl_to_symbol$Ensembl_ID)]
  df <-as.data.frame(cbind(result_table_sorted, Gene_name = symbol_list))
  
  # Write sorted table to file
  write.table(df, file = paste0("./DE/",file_name,".txt"), 
            sep = "\t", col.names=NA)
  return(df)
}
```

```{r}
# Calculate DE for all_data samples
#design(dds.rnaseA) <- ~Treatment # Removid Read.depth from formula given that all samples are Read.depth = High

dds_res$group <- relevel(dds_res$group, "37oC_WT")
dds_res <- DESeq(dds_res)
my_contrasts <- resultsNames(dds_res)


# Using lfcShrink instead of results to reduce high Log2FC bias of genes with low expression
# 37oC mutant vs WT
res_mut_vs_WT_37C <- lfcShrink(dds_res, coef = my_contrasts[4], type = "ashr", )
res_mut_vs_WT_23C <- lfcShrink(dds_res, contrast = c("group", "23oC_mutant", "23oC_WT"), type = "ashr", )
# Another way to get the same reult as res_mut_vs_WT_23C
#res_mut_vs_WT_23C_2 <- lfcShrink(dds_res, contrast = c(0, 1, -1, 0), type = "ashr", )

summary(res_genotype_SD105_vs_WT, alpha = 0.05)
summary(res_genotype_SD106_vs_WT, alpha = 0.05)
summary(res_genotype_SD107_vs_WT, alpha = 0.05)

# Sort results by Log2FC
res_genotype_SD105_vs_WT_sorted <- sort_and_write_res_table(result_table = res_genotype_SD105_vs_WT, file_name = paste0("DE_SD105_vs_WT")) 
res_genotype_SD106_vs_WT_sorted <- sort_and_write_res_table(result_table = res_genotype_SD106_vs_WT, file_name = paste0("DE_SD106_vs_WT")) 
res_genotype_SD107_vs_WT_sorted <- sort_and_write_res_table(result_table = res_genotype_SD107_vs_WT, file_name = paste0("DE_SD107_vs_WT")) 



```
```{r}
table_counts_normalized <- counts(dds_res, normalized=TRUE)
write.table(x = as.data.frame(table_counts_normalized), file = "read_counts_deseq2_normalized.txt", sep = "\t", col.names = NA)
```
  

```{r}
print(sessionInfo())
```